Implement M* as an SIA option in addition to our Mahaffy schemes

bueler opened the pull request:
This draft PR is, for now, intended to start a conversation.

See https://doi.org/10.1017/jog.2015.3 for the paper where the M* scheme for the SIA is described.  I hope that the scheme may remove some of the high-frequency chatter (checkerboard) seen in e.g. Thomas Franks' SIA results.  

Our existing schemes for SIA fluxes all arise from Mahaffy, though different surface gradient calculations are implemented.  Note there are two advantages of M*: better quadrature, upwinding of the advective part of the SIA flux (when the bed is nonflat).  I am starting by just trying to implement the better quadrature.

**Checklist**

- [ ] update documentation
- [ ] update [`CHANGES.rst`](CHANGES.rst)
- [ ] add regression tests

bueler committed:
Start implementation of M* method for SIA.

This compiles.  It abuses the -gradient option by tacking on one
more case ("-gradient mstar") and then using it to compute the
diffusive flux (SIAFD::compute_diffusive_flux_mstar()).
It abuses the flow law mechanism by directly evaluating the SIA
formula.

A bunch of stuff is not yet implemented; see FIXMEs in
compute_diffusive_flux_mstar().

bueler commented on src/stressbalance/sia/SIAFD.cc:
```diff
@@ -799,6 +804,63 @@ void SIAFD::compute_diffusive_flux(const IceModelVec2Stag &h_x, const IceModelVe
+        //   while _left,_right are 2,1
+        double qint = 0.0; // estimate of integral of flux along face
+        if (o == 0) {
+          // next lines use equation (18) in Bueler 2016
```
@ckhroulev Advise on how to use the existing Q1 FEM tools here?

ckhroulev replied:
See 1efe3ab217c64f4158023e94b0aeb606b4b7a650 for one way of using existing FE tools.

bueler commented on src/stressbalance/sia/SIAFD.cc:
```diff
@@ -799,6 +804,63 @@ void SIAFD::compute_diffusive_flux(const IceModelVec2Stag &h_x, const IceModelVe
   } // o-loop
 }
 
+double SIAFD::q_mstar(const double H, const double slope, const double gradscomponent) {
```
@ckhroulev What's the easiest way to get the existing flow law to compute this?  I just want isothermal SIA for testing, but the other methods see a vertical grid and they have the resulting artifacts from finite Mz, so I know I won't be comparing apples to apples if I get this mstar stuff working this way.

ckhroulev replied:
I'm not sure yet. If we get serious about this in PISM we'd have to (this is probably an incomplete list)

- evaluate `q_mstar()` in 3D using ice softness from the flow law and average the flux
- compute diffusivity and its byproducts needed to compute the 3D velocity *or* recover diffusivity from the flux (using surface slope) and re-write the 3D velocity computation
- add a bit of code to handle ice margins (this would replace the `haseloff` gradient method)

bueler committed:
Use util/fem/FEM.hh tool chi() to draft initial M* implementation.

bueler committed:
Caught some bugs.  Now result m_diffusive_flux is 10^10 too large.

ckhroulev committed:
Use existing FE tools

ckhroulev committed:
Fix ice softness

ckhroulev committed:
Add a FIXME comment

ckhroulev commented on src/stressbalance/sia/SIAFD.cc:
```diff
@@ -799,6 +804,63 @@ void SIAFD::compute_diffusive_flux(const IceModelVec2Stag &h_x, const IceModelVe
+  const double n   = m_flow_law->exponent(), // presumably 3.0
+               g   = m_config->get_number("constants.standard_gravity"),
+               rho = m_config->get_number("constants.ice.density"),
+               A   = m_config->get_number("flow_law.Paterson_Budd.A_cold"),
```
This does not match the value used in `pismv`, which explains the difference in the magnitude of the flux (see 07002d94eea2ae156052e86cf4f352beb388819a).

ckhroulev committed:
Remove debugging code

bueler committed:
Fix typo.  Add back debugging output because it found this typo.

Before this commit there is a visible asymmetry in output diffusive_flux.nc.

mstar reduces Mahaffy maximum errors on Test B.  Do
   pismv -test B -ys 422.45 -y 1000 -gradient X -o X.nc
for X = mahaffy, haseloff, mstar.

mahaffy = haseloff:
NUMERICAL ERRORS evaluated at final time (relative to exact solution):
geometry  :    prcntVOL        maxH         avH   relmaxETA
               0.080245  198.584456    4.566908    0.011726

mstar:
NUMERICAL ERRORS evaluated at final time (relative to exact solution):
geometry  :    prcntVOL        maxH         avH   relmaxETA
               0.080245  189.647758    4.900889    0.011708

ckhroulev commented on src/stressbalance/sia/SIAFD.cc:
```diff
@@ -923,12 +925,12 @@ void SIAFD::compute_diffusive_flux_mstar(const Geometry &geometry,
-        NE.reset(i - 1, j);
-        NE.nodal_values(ice_thickness, H);
-        NE.nodal_values(ice_surface, S);
+        NW.reset(i - 1, j);
```
Ah! Thanks.

bueler committed:
Sorry.  Now that it has done its duty, you are right it should be removed.

ckhroulev commented on src/stressbalance/sia/SIAFD.cc:
```diff
@@ -160,8 +160,6 @@ void SIAFD::update(const IceModelVec2V &sliding_velocity,
   }
   profiling.end("sia.flux");
 
-  m_diffusive_flux.dump("diffusive_flux.nc");  // DEBUG
```
Now that it no longer crashes we can save `flux_staggered` to `-extra_file`, or set `-output.sizes.medium flux_staggered` to save it to output files.

bueler replied:
:smile:  Cool.  `flux_staggered` is the unfiltered output of `SIAFD::update()`?  I was not sure if it was modified later.

ckhroulev replied:
Well, no. Now that I looked at the code I see that it *is* modified.
- It includes sliding (if present).
- It does not include fluxed between cells considered "ice free" according to PISM's cell type mask.

But in this case (no sliding) it is almost the same as SIAFD's diffusive flux, and I think it would be the same with `geometry.ice_free_thickness_standard` set to zero.

ckhroulev commented:
> Just curious: This is the same as `double q7 = 0.0;`?

Yes, but using somewhat more modern syntax which is consistent across different data types (classes and fundamental types).

ckhroulev commented:
> Just curious: Why the explicit scoping? Should `SE` declaration be put inside?

Regarding scoping: no particularly good reason. I just like to be able to see that a bit of code is a "unit" that sets the variable defined right above it. Sometimes I use code blocks the way you would use paragraphs when writing prose.

Regarding `SE`, etc: no. Its constructor pre-computes values of shape functions at quadrature point locations, etc. We don't want to repeat that at each grid point.

bueler commented:
Cool.  I would not have 

> > Just curious: Why the explicit scoping? Should `SE` declaration be put inside?
> 
> Regarding scoping: no particularly good reason. I just like to be able to see that a bit of code is a "unit" that sets the variable defined right above it. Sometimes I use code blocks the way you would use paragraphs when writing prose.
> 
> Regarding `SE`, etc: no. Its constructor pre-computes values of shape functions at quadrature point locations, etc. We don't want to repeat that at each grid point.

Got it.

bueler committed:
Now we compute the split form flux (q=-DgradH+WH^n+2), but w/o upwinding.

bueler committed:
Fully implement M* scheme, including upwinding.

This should be the full M* scheme for the isothermal SIA, i.e. this is the
explicit time-stepping version of the scheme in Bueler 2016.  However,
it is only implemented for the isothermal SIA, and it is badly implemented
because of code duplication and known-to-be slow choices.  (See FIXMEs in
W_mstar() and q_mstar().)  No non-flat bed testing has yet been done, but it
compiles and produces the same numbers as the previous M* implementation on
flat-bed Test B.
